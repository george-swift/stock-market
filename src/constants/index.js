export const API_BASE_URL = 'https://financialmodelingprep.com/api/v3/';

export const FETCH_ETF_LIST_STARTED = 'FETCH_ETF_LIST_STARTED';

export const FETCH_ETF_LIST_FAILED = 'FETCH_ETF_LIST_FAILED';

export const FETCH_ETF_LIST_SUCCEEDED = 'FETCH_ETF_LIST_SUCCEEDED';

export const FILTER_LISTING = 'FILTER_LISTING';

export const FILTER_NASDAQ = 'FILTER_NASDAQ';

export const FETCH_NASDAQ_STARTED = 'FETCH_NASDAQ_STARTED';

export const FETCH_NASDAQ_FAILED = 'FETCH_NASDAQ_FAILED';

export const FETCH_NASDAQ_SUCCEEDED = 'FETCH_NASDAQ_SUCCEEDED';

export const UI = { isLoading: false, error: null };

export const FILTER = { listing: '', nasdaq: '' };

export const sortByValue = (data) => data.sort((item1, item2) => item2.price - item1.price);
