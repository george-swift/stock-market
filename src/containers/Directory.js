import { useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { FaRegFolderOpen, FaSpinner } from 'react-icons/fa';
import { filterListing, filterMarkets } from '../actions';
import { getFilteredList } from '../selectors';
import FlashMessage from '../components/FlashMessage';
import ExchangeFilter from '../components/ExchangeFilter';
import List from '../components/List';

const Directory = () => {
  const { isLoading, error } = useSelector((state) => state.notifications);
  const companies = useSelector((state) => getFilteredList(state));
  const dispatch = useDispatch();

  const handleMarkets = (filter) => dispatch(filterMarkets(filter));

  const handleSearch = (e) => dispatch(filterListing(e.target.value));

  const navigate = useNavigate();

  const getUrl = (symbol) => navigate(`/company/${symbol}`);

  return (
    <>
      <h2>
        <FaRegFolderOpen />
        <span className="ms-3">Listing directory</span>
      </h2>
      {
        error !== null && (
          <FlashMessage>
            Error...Try reloading the page.
          </FlashMessage>
        )
      }
      {isLoading && <p className="page-loading"><FaSpinner /></p>}
      <div>
        <ExchangeFilter filterExchange={handleMarkets} />
        <input
          type="search"
          className="form-control form-control-sm mb-3"
          placeholder="Search trading symbol e.g GE"
          onChange={handleSearch}
        />
        <ul className="p-0 listings">
          {
            companies?.map((company) => (
              <List
                key={company.symbol}
                symbol={company.symbol}
                name={company.name}
                exchange={company.exchange}
                getUrl={getUrl}
              />
            )) ?? <li>No companies listed in directory</li>
          }
        </ul>
      </div>
    </>
  );
};

export default Directory;
