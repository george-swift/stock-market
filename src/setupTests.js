/* eslint import/no-extraneous-dependencies: ["error", {"peerDependencies": true}] */
import { configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });
