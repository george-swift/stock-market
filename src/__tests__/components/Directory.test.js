import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import '../../setupTests';
import { UI, FILTER } from '../../constants';
import Directory from '../../containers/Directory';

const mockStore = configureStore();

describe('Snapshot of the Directory component', () => {
  const store = mockStore({
    listings: [
      {
        symbol: 'MU',
        name: 'Micron Technology, Inc.',
        price: 75.53,
        exchange: 'Nasdaq Global Select',
      },
      {
        symbol: 'LOW',
        name: "Lowe's Companies, Inc.",
        price: 196.77,
        exchange: 'New York Stock Exchange',
      },
    ],
    filter: { ...FILTER },
    notifications: { ...UI },
  });

  it('should render correctly', () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <Directory />
        </Router>
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('List').exists()).toBe(true);
  });
});

describe('Querying the Directory component', () => {
  it('should render a spinning component while the page loads', () => {
    const store = mockStore({
      listings: [],
      filter: { ...FILTER },
      notifications: {
        ...UI,
        isLoading: true,
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <Directory />
        </Router>
      </Provider>,
    );
    expect(wrapper.find('FlashMessage').exists()).toBe(false);
    expect(wrapper.find('FaSpinner').exists()).toBe(true);
    expect(wrapper.find('List').exists()).toBe(false);
  });

  it('should render an alert message if there is an error', () => {
    const store = mockStore({
      listings: [],
      filter: { ...FILTER },
      notifications: {
        ...UI,
        isLoading: false,
        error: 'Error loading directory',
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <Directory />
        </Router>
      </Provider>,
    );

    expect(wrapper.find('FlashMessage').exists()).toBe(true);
    expect(wrapper.find('FaSpinner').exists()).toBe(false);
    expect(wrapper.find('List').exists()).toBe(false);
  });
});
