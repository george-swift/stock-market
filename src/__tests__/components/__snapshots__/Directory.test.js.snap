// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of the Directory component should render correctly 1`] = `ReactWrapper {}`;
